### PR TITLE
`when`/`case` + general cleanup

### PR DESCRIPTION
**Pull request by drkameleon**
# Description

Continuing from here: https://github.com/arturo-lang/arturo/pull/1653

## Type of change

- [x] Code cleanup
- [x] Enhancement (implementation update, or general performance enhancements)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation (documentation-related additions)

### CODE REVIEW

**Comment by drkameleon** [comment]
⚠️ **examples/rosetta/read a configuration file.art**

Unit-test failing!

File: examples/_drafts/parser.art [hunk 1]
```diff
@@ -37,8 +37,8 @@ parse: function [in,rules][
     while [i < size rules][
         rule: rules \ i
         item: in \ pos
-        case [(type rule)=]
-            when? [:type][
+        when type rule [
```
[the remark this answers is not in the document]
Comment (drkameleon) [reply examples/_drafts/parser.art:40]: that's quite messy indeed. I have to go through the existing code again and make sure nothing is missing.

Also, I'm not even sure I touched the `_drafts` folder...

File: examples/rosetta/bioinformatics - base count.art [hunk 2]
```diff
@@ -19,12 +19,12 @@ prettyPrint: function [in][
         print split.every:10 line
 
         loop split line 'ch [
```
Comment (Eloitor) [inline examples/rosetta/bioinformatics - base count.art:21]: tally

File: examples/rosetta/rot-13.art [hunk 3]
```diff
@@ -1,8 +1,10 @@
 rot13: function [c][
-    case [in? lower c]
-        when? -> 'a'..'m' -> return to :char (to :integer c) + 13
-        when? -> 'n'..'z' -> return to :char (to :integer c) - 13
-        else  -> return c
+    lc: lower c
+    when [
+        in? lc 'a'..'m' -> return to :char (to :integer c) + 13
```
Comment (Eloitor) [inline examples/rosetta/rot-13.art:4]: between?

File: examples/rosetta/bioinformatics - base count.art (hunk 2 above)
Comment (drkameleon) [reply examples/rosetta/bioinformatics - base count.art]: That was... brutal. Practically half of the code is gone! 🤣

**Comment by drkameleon** [comment]
## Final version

(nothing written in stone; but to avoid going over it till the end of time and not merging it, let's merge it and we can change it as we go, if we need to do so. also, the approach is quite simplistic, less intrusive than initially designed, so that we don't overcomplicate things too much too soon; we have all the time in the world for that... lol)

### `case`

> match given argument against different values and execute corresponding block

```red
x: 2

; the main block is always evaluated!
case x [
    1   -> print "x is one!"
    2   -> print "x is two!"
    any -> print "x is none of the above"
]
; x is two!
```
```red
key: "one"
case key [
    "one" 1,        ; we can also return
    "two" 2         ; simple constant values directly
]
; => 2
```
```red
case "hello" #[
    hello: "hola"
    adios: "goodbye"
]
; => "hola"
```

### `when`

> check conditions one by one and execute corresponding block accordingly

```red
; the main block is always evaluated!
when [
    prime? 4 -> print "yes, 4 is prime - wait, what?!"
    prime? 5 -> print "yes, 5 is prime
    prime? 7 -> print "yes, 6 is prime
    true     -> print "none of the above was true"
]
; yes, 5 is prime
```
```red
when.any [
    prime? 4 -> print "yes, 4 is prime - wait, what?!"
    prime? 5 -> print "yes, 5 is prime"
    prime? 7 -> print "yes, 7 is prime"
]
; yes, 5 is prime
; yes, 7 is prime
```
```red
x: 2
when.has: x [
    [=0] -> print "x is zero!"
    [<1] -> print "x is less than 1"
    [<4] -> print "x is less than 4"
    true -> print "x is >= 4"
]
; x is less than 4
```

**Comment by drkameleon** [comment]
And...
ready to merge! 🚀

**Pull request merged**